python/mxnet/module/base_module.py

```diff
@@ -523,8 +524,9 @@ def fit(self, train_data, eval_data=None, eval_metric='acc',
                     monitor.toc_print()
 
                 if batch_end_callback is not None:
+                    arg_eval_metric = eval_metric if not end_of_batch else deepcopy(eval_metric)
```

Comment:
This is too hacky and potentially very slow

Reply:
Deep copy happens maximum once a epoch. That adds negligible time to an epoch.

Hacky - yes. That's because we are trying to print metrics outside the loop which could get cleared inside the loop. Any solution that doesn't modify the existing behavior will be a little hacky.

Reply:
If you can think of a less hacky solution, please propose. If not, I think this is better than printing accuracy as nan in training log which creates bad user experience. A number of users have reported this.